### PR TITLE
Update Set-AzureRmOsDisk cmdlet for Disk Encryption

### PR DESCRIPTION
**Pull request by hyonholee**

### CODE REVIEW

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs [hunk 1]
```diff
@@ -97,8 +105,69 @@ public class SetAzureVMOSDiskCommand : Microsoft.Azure.Commands.ResourceManager.
...
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = HelpMessages.VMOSDiskLinuxOSType)]
         public SwitchParameter Linux { get; set; }
 
+        [Parameter(
+            ParameterSetName = WindowsDiskEncryptionParameterSet,
+            Mandatory = true,
+            Position = 7,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = HelpMessages.VMOSDiskLinuxOSType)]
```
Comment (huangpf) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs]: `VMOSDiskWindowsOSType`   ?

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs [hunk 2]
```diff
@@ -97,8 +105,69 @@ public class SetAzureVMOSDiskCommand : Microsoft.Azure.Commands.ResourceManager.
...
+        public string DiskEncryptionKeyVaultId { get; set; }
+
+        [Parameter(
+            ParameterSetName = WindowsDiskEncryptionParameterSet,
+            Mandatory = false,
+            Position = 9,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = HelpMessages.VMOSDiskLinuxOSType)]
```
Comment (huangpf) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs]: same

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs [hunk 3]
```diff
@@ -97,8 +105,69 @@ public class SetAzureVMOSDiskCommand : Microsoft.Azure.Commands.ResourceManager.
...
+        public string KeyEncryptionKeyUrl { get; set; }
+
+        [Parameter(
+            ParameterSetName = WindowsDiskEncryptionParameterSet,
+            Mandatory = false,
+            Position = 10,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = HelpMessages.VMOSDiskLinuxOSType)]
```
Comment (huangpf) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs]: same

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs [hunk 4]
```diff
@@ -106,6 +175,14 @@ protected override void ProcessRecord()
                 this.VM.StorageProfile = new StorageProfile();
             }
 
+            if ((this.KeyEncryptionKeyVaultId == null && this.KeyEncryptionKeyUrl != null)
+                || (this.KeyEncryptionKeyVaultId != null && this.KeyEncryptionKeyUrl == null))
```
Comment (huangpf) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs]: `string.IsNullOrEmpty`  ?

**Comment by huangpf** [comment]
LGTM

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs [hunk 5]
```diff
@@ -97,8 +105,69 @@ public class SetAzureVMOSDiskCommand : Microsoft.Azure.Commands.ResourceManager.
...
+            ParameterSetName = LinuxDiskEncryptionParameterSet,
+            Mandatory = true,
+            Position = 8,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = HelpMessages.VMOSDiskLinuxOSType)]
+        public string DiskEncryptionKeyVaultId { get; set; }
+
+        [Parameter(
+            ParameterSetName = WindowsDiskEncryptionParameterSet,
+            Mandatory = false,
```
Comment (huangpf) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs:145]: You could probably mandate this in a separate parameter set, i.e. `WindowsDiskAndKeyEncryptionParameterSet`.

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs [hunk 6]
```diff
@@ -119,7 +196,31 @@ protected override void ProcessRecord()
...
+                    : new KeyVaultKeyReference
+                    {
+                        KeyUrl = this.KeyEncryptionKeyUrl,
+                        SourceVault = new SourceVaultReference
+                        {
+                            ReferenceUri = this.KeyEncryptionKeyVaultId
+                        },
+                    }
+                }
+                : null
```
Comment (huangpf) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs:223]: How about linux?

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs [hunk 7]
```diff
@@ -119,7 +196,31 @@ protected override void ProcessRecord()
                 {
                     Uri = this.SourceImageUri
                 },
-                CreateOption = this.CreateOption
+                CreateOption = this.CreateOption,
+                EncryptionSettings =
+                (this.ParameterSetName.Equals(WindowsDiskEncryptionParameterSet) || this.ParameterSetName.Equals(WindowsDiskEncryptionParameterSet))
```
Comment (huangpf) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs]: It's a bug that one of them should be `Linux` .

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs [hunk 8]
```diff
@@ -119,7 +196,31 @@ protected override void ProcessRecord()
...
+                ? new DiskEncryptionSettings
+                {
+                    DiskEncryptionKey = new KeyVaultSecretReference
+                    {
+                        SourceVault = new SourceVaultReference
+                        {
+                            ReferenceUri = this.DiskEncryptionKeyVaultId
+                        },
+                        SecretUrl = this.DiskEncryptionKeyUrl
```
Comment (huangpf) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs:210]: Referring to the client field, the PS parameter probably should be `DiskEncryptionSecretUrl`.

Comment (AzureRT) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs:210]: These parameter names are decided after discussion.  If we follow the client field, it should be DiskEncryptionKeySecretUrl, but it feels redundunt, so removed 'Secret'.

Comment (huangpf) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/SetAzureVMOSDiskCommand.cs:210]: OK. Thanks.

**Comment by huangpf** [comment]
LGTM.

**Pull request merged**